rfd/XXXX-teleport-release-engineering.md

```diff
@@ -0,0 +1,320 @@
+    - Dockerfile sources for OCI Container Images
+    - Sources for the Teleport PAM modules
+    - Sources for various packaging tools (Debian, RPM, macOS)
+    - Unix shells scripts for automating build tasks
+    - Powershell scripts for building Teleport on Windows
+  - `./examples` - indended to show examples of configuring and running Teleport,
+    some examples serve double duty
+    - `etcd` keys and config used during CI
+    - `charts` providing sources for publiushed helm charts
+    - Scripts and resources for testing Kubernetes cluster
+
+#### **Infrastructure**
+
+Drone is entirely self hosted, we (Teleport) need to manage it ourselves. 
+
+GHA, on the other hand, is largely managed for us by GitHub, although we
+do use pools of ephemeral, self-hosted runners for specific tasks. 
+
+We currently manage, as part of our CI/CD/Release system:
+ - AWS EKS cluster running `Drone` (see `./drone` in the `ops` repository)
+ - **Drone runners:**
+    - 3x (one per actively supported major Teleport Version) AWS EC2 AMD64
+      instances for Windows Builds
+    - Intel MacMini in Macstadium
+ - **Self-Hosted GHA Runners**
+    - Ephemeral AWS EC2 arm64 boxes for arm64 linux builds. Pool has a max of 10
+    instances
+    - Ephemeral EC2 instance with a persistent volume for building `deb` 
+    and `rpm` packages.
+    
+### Goals of a Release Process
+
+At a a high level, we want a release process that is:  
+  * *Consistent* - we want to provide a stable framwork for delivering 
+    releases reliably from source,
+  * *Agile* - we want to be able to quickly integrate new delvery channels
+    or products without having to redesign everything. 
+  * *Identifiable* - we can identify all the tools and components used
+    to deliver a release, in order to mitigate supply chain attacks
+
+We also want some functional requirements:
+
+ 1. An engineer *must* be able to at least construct a Teleport binary at their 
+    desk with a minimum of fuss.
+ 2. An engineer *must* be able to run close to 100% of all tests at their desk
+    with a minimum of fuss 
+ 3. The release envirionment should be as close as practical to the CI 
+    environment, ideally the same. We don't want surprises come release time. 
+ 4. An engineer should be able to cut a debug or testing release at any time 
+    using the proper release procedure and tooling. Especially if that release
+    will travel outside the Teleport organisation.
+
+Non-functional requirements:
+
+### Suggestions
+
+#### **Draw some boundaries**
+
+One cause of the complexity in our build system is that we have not (in the
+last 2 years at least) collectively sat down and thought about what we want
+our Delivery system to *be*. What is its *purpose*, and what boundaries can
+we put on it.
+
+Some open questions are:
+
+ - What tasks should an engineer be able to do at their desk?
+   - Build binaries
+   - Run Tests
+ - Where is the boundary between
+   - Desk/CI
+   - Desk/Delivery
```

Comment:
What would "unsigned release artifacts" contain in this case? Is it just the first-order tarballs, or do you include docker images, Debian packages, AMIs, etc?

Reply:
> What would "unsigned release artifacts" contain in this case? 

The dividing line to me is "what can be built on a single machine".  Native binaries are out. AMIs are out.

*  first-order binaries, yes
*  release tarball (or their moral equivalent on the target platform), yes
*  docker images, yes
*  debian packages, yes
*  AMIs - no -- this would require aws access and runnning stuff off the machine